packages/common/src/sendTransaction.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  Account,
+  CallParameters,
+  Chain,
+  Client,
+  SendTransactionParameters,
+  Transport,
+  WriteContractReturnType,
+} from "viem";
+import { call, sendTransaction as viem_sendTransaction } from "viem/actions";
+import pRetry from "p-retry";
+import { debug as parentDebug } from "./debug";
+import { getNonceManager } from "./getNonceManager";
+import { parseAccount } from "viem/accounts";
+
+const debug = parentDebug.extend("sendTransaction");
+
+// TODO: migrate away from this approach once we can hook into viem's nonce management: https://github.com/wagmi-dev/viem/discussions/1230
+
+export async function sendTransaction<
+  TChain extends Chain | undefined,
+  TAccount extends Account | undefined,
+  TChainOverride extends Chain | undefined
+>(
+  client: Client<Transport, TChain, TAccount>,
+  request: SendTransactionParameters<TChain, TAccount, TChainOverride>
+): Promise<WriteContractReturnType> {
+  const account_ = request.account ?? client.account;
+  if (!account_) {
+    // TODO: replace with viem AccountNotFoundError once its exported
+    throw new Error("No account provided");
+  }
+  const account = parseAccount(account_);
+
+  const nonceManager = await getNonceManager({
+    client,
+    address: account.address,
+  });
+
+  async function prepare(): Promise<SendTransactionParameters<TChain, TAccount, TChainOverride>> {
+    if (request.gas) {
+      debug("gas provided, skipping simulate", request.to);
+      return request;
+    }
+
+    debug("simulating tx to", request.to);
+    await call(client, {
+      ...request,
+      blockTag: "pending",
+      account,
+    } as CallParameters<TChain>);
+
+    // TODO: estimate gas
+
+    return request;
+  }
+
+  const preparedRequest = await prepare();
+
+  return await nonceManager.mempoolQueue.add(
+    () =>
+      pRetry(
+        async () => {
+          if (!nonceManager.hasNonce()) {
+            await nonceManager.resetNonce();
+          }
+
+          const nonce = nonceManager.nextNonce();
+          debug("sending tx with nonce", nonce, "to", preparedRequest.to);
+          return await viem_sendTransaction(client, { nonce, ...preparedRequest });
+        },
+        {
+          retries: 3,
+          onFailedAttempt: async (error) => {
+            // On nonce errors, reset the nonce and retry
+            if (nonceManager.shouldResetNonce(error)) {
+              debug("got nonce error, retrying", error.message);
+              await nonceManager.resetNonce();
+              return;
+            }
+            // TODO: prepare again if there are gas errors?
+            throw error;
+          },
+        }
+      ),
+    { throwOnTimeout: true }
+  );
```

Comment:
>  with mempoolQueue concurrency set to 1 we wait for the transaction to be accepted by the rpc node before attempting to send the next transaction, but we're not waiting for the transaction to be included in a block, so it's still possible to include multiple transactions in one block, correct?

correct

> While before we were "preparing" to see if the node would accept the transaction, but we weren't actually waiting for it to accept it before sending another one, so in some edge case it could have happened that we're sending two transactions with subsequent nonces, the first one gets dropped for some reason, and the second one is stuck?

correct! or, what I think I was seeing is that when parallelizing lots of txs, the node would receive the set of txs out of nonce order, so some would just fail/not be accepted by the node